rusoto/core/src/event_stream.rs

```diff
@@ -0,0 +1,491 @@
+
+#[doc(hidden)]
+pub trait DeserializeEvent: Sized {
+    fn deserialize_event(event_type: &str, data: &[u8]) -> Result<Self, RusotoError<()>>;
+}
+
+#[derive(Debug, Eq, PartialEq)]
+enum EventStreamParseError {
+    UnexpectedEof,
+    InvalidCrc,
+    InvalidData(&'static str),
+}
+
+fn check_crc32(data: &[u8], ref_value: u32) -> Result<(), EventStreamParseError> {
+    let mut hasher = Hasher::new();
+    hasher.update(data);
+
+    if hasher.finalize() != ref_value {
+        Err(EventStreamParseError::InvalidCrc)
+    } else {
+        Ok(())
+    }
+}
+
+fn read_slice<'a>(reader: &mut &'a [u8], size: usize) -> Result<&'a [u8], EventStreamParseError> {
+    if reader.len() < size {
+        return Err(EventStreamParseError::UnexpectedEof);
+    }
+
+    let slice = &reader[..size];
+    *reader = &reader[size..];
+    Ok(slice)
+}
+
+fn read_u8(reader: &mut &[u8]) -> Result<u8, EventStreamParseError> {
+    let buf = read_slice(reader, std::mem::size_of::<u8>())?
+        .try_into()
+        .unwrap();
+    Ok(u8::from_be_bytes(buf))
+}
+
+fn read_u16(reader: &mut &[u8]) -> Result<u16, EventStreamParseError> {
+    let buf = read_slice(reader, std::mem::size_of::<u16>())?
+        .try_into()
+        .unwrap();
+    Ok(u16::from_be_bytes(buf))
+}
+
+fn read_u32(reader: &mut &[u8]) -> Result<u32, EventStreamParseError> {
+    let buf = read_slice(reader, std::mem::size_of::<u32>())?
+        .try_into()
+        .unwrap();
+    Ok(u32::from_be_bytes(buf))
+}
+
+fn read_u64(reader: &mut &[u8]) -> Result<u64, EventStreamParseError> {
+    let buf = read_slice(reader, std::mem::size_of::<u64>())?
+        .try_into()
+        .unwrap();
+    Ok(u64::from_be_bytes(buf))
+}
+
+impl EventStreamParseError {
+    fn eof_as_invalid(self) -> Self {
+        match self {
+            EventStreamParseError::UnexpectedEof => {
+                EventStreamParseError::InvalidData("Malformed event: ended unexpectedly")
+            }
+            other => other,
+        }
+    }
+}
+
+impl std::error::Error for EventStreamParseError {}
+
+impl Display for EventStreamParseError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            EventStreamParseError::UnexpectedEof => write!(f, "Expected additional data"),
+            EventStreamParseError::InvalidCrc => write!(f, "CRC check failed"),
+            EventStreamParseError::InvalidData(msg) => write!(f, "{}", msg),
+        }
+    }
+}
+
+impl<T> Into<RusotoError<T>> for EventStreamParseError {
+    fn into(self) -> RusotoError<T> {
+        RusotoError::ParseError(self.to_string())
+    }
+}
+
+#[allow(missing_docs)]
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+enum EventStreamHeaderValue<'a> {
+    Bool(bool),
+    UInt8(u8),
+    UInt16(u16),
+    UInt32(u32),
+    UInt64(u64),
+    ByteArray(&'a [u8]),
+    String(&'a str),
+    Timestamp(u64),
+    Uuid(&'a [u8; 16]), // don't want to pull the uuid dependency for this, so just u8
```

Comment:
wish the `uuid` crate supported `From<&'a [u8; 16]>` but this is definitely the right type to have here :(